### DISCUSSION TITLE
Pattern recognition / color threshold pixel counter

### DISCUSSION THREAD

**Discussion by DynV** [ImageMagick/ImageMagick#2490] [Help] [answered]
I have hundreds of screenshots I'd like to pick some images from, which have a pattern. Here's an example
![FO4-Loriya-1st_perk-Ghoulish_(L80)-cropped_center_dialog](https://user-images.githubusercontent.com/2823807/91765794-8fa14c80-eba7-11ea-8c27-bb4a28bae9ba.png)
and here's the pattern I see from it
![FO4-Loriya-1st_perk-Ghoulish_(L80)-cropped_center_dialog-color_picked_center_1st_star_threshold_25](https://user-images.githubusercontent.com/2823807/91765837-a34cb300-eba7-11ea-82c7-462f6c98bfa8.png)
which I took using GIMP color picker with a threshold of 25 and clicking pretty close to the center of the leftmost star (content of the selection filled with red to show the selected area). Those are cropped and here's a shrunk version of the full screenshot with the color picked
![FO4-Loriya-1st_perk-Ghoulish_(L80)-color_picked_center_1st_star_threshold_25-sz_25pc-70pc](https://user-images.githubusercontent.com/2823807/91766388-819ffb80-eba8-11ea-814f-b9e1e74668fd.jpg)

As one may see, there's plenty selected in the screenshot, and if there would be a way to get all the colors of that color threshold and count them, the screenshots I'm looking for it would most likely stand out. The information could be gathered in a flat-file DB looking something like
File name; Color 1 (pixel count); Color 2; Color 3...
SS1.jpg; 5; 10; 3; 12
SS2.jpg; 100; 80; 90; 70
SS3.jpg; 13; 6; 5; 22
and from the above, SS2.jpg would then be pre-selected, and if the user confirmed a match, it would be added to a list manually.

The above paragraph is just an example, and if there's a way to make a pre-selection with little-to-no false negative and not a ton of false positives, I'd be happy.

Thank you kindly

**Update 1:**

The dialog window moves (from screenshot to screenshot) but not the bottom interface, although in that there's changes, the # and the thin horizontal rectangle just below is a meter (filling from the left).

**Comment by fmw42** [comment]
I do not understand your question. What is your starting image and what resulting image do you need.  If you want to color thresholding, you can use -fuzz XX% +transparent "rgb(r,g,b)". That will make everything else but the color transparent. The fuzz XX% will allow some variation in color. A pure color would use -fuzz 0%. If you want to know the colors, then get the histogram and ignore the transparent ones.

**Comment by DynV** [comment]
I need to recognize a pattern in a bunch of screenshots, not 1, many. The way I'm thinking would be easiest, would be to count the # of pixels for a set # of colors, for all the screenshots and to have the data listed somewhere (I suggested in a flat-file DB).

**Comment by fmw42** [comment]
If the pattern is the same size and orientation in all images, then you can use the compare tool in ImageMagick. See 

https://imagemagick.org/script/compare.php
https://imagemagick.org/Usage/compare/

**Comment by DynV** [comment]
That seems nice, but there's hundreds of screenshots to process. I'm assuming many of the comparison results would be plain black. Is there a way to get a list of pixel count that would indicate if there's likely a match? ie: a list of the images and its black pixels count.

**Reply by fmw42** [reply]
Pick your metric and use null: for the output. The AE metric is just a count of pixels where there is a difference, the rmse metric is the root mean squared difference, which is like an average difference. Those are the two most popular metrics. See the list of metrics at https://imagemagick.org/Usage/compare/

**Comment by DynV** [comment]
I'm not sure if I'm doing it right but I made a sample of 4 screenshots, the 2 middle close to what I want, the 3rd, being an exact match I'm looking for; so it would be something like this: 1 not even close, 2 pretty close, 3 pretty darn close / exact, 4 close to nothing matches. The comparison is basically all the bright green from the 3rd screenshot minus the # and some other part that might be different from other matches on a transparent background, which is
![FO4-new_perk_shared_content](https://user-images.githubusercontent.com/2823807/91781721-013dc280-ebc9-11ea-851d-1a99015b5884.png)

Here's the censored command output:
> USER_NAME@SYSTEM_NAME ~
> $ compare -metric AE FO4-new_perk_shared_content.png SS/SS1.jpg null: 2>&1
> 8.29439e+06
> USER_NAME@SYSTEM_NAME ~
> $ compare -metric AE FO4-new_perk_shared_content.png SS/SS2.jpg null: 2>&1
> 8.28862e+06
> USER_NAME@SYSTEM_NAME ~
> $ compare -metric AE FO4-new_perk_shared_content.png SS/SS3.jpg null: 2>&1
> 8.28542e+06
> USER_NAME@SYSTEM_NAME ~
> $ compare -metric AE FO4-new_perk_shared_content.png SS/SS4.jpg null: 2>&1
> 8.29414e+06

Did I do it properly? If so, how should I interpret the data? Is there a way to make it easier to spot how close of a match it is?

**Reply by fmw42** [reply]
It is very hard to access how many images you have attached since they are transparent. Would you zip them and attach the zip file. What is your question? What are you trying to do? You are getting very small values because most of your image is transparent. Perhaps you should just crop out the region where it has something other than transparent and compare the cropped regions.  I took two of your images, did a virtual trim to get the region where not transparent and cropped them, but get 0 differences.

```
cropvals=`convert 1.png -format %@ info:`
compare -metric AE 1.png[$cropvals] 2.png[$cropvals] null:
0
```

What are you expecting?

**Reply by DynV** [reply]
In my previous post, there's only 1 image attached, only what's not transparent is to be analyzed for a match. Since the images to be analyzed have a non-negligible size and I'm not sure if there's a cap on github, or this particular one, I've uploaded them [here](https://imgur.com/a/Je4Ppp9); please notice in their description is the file name I'm using for this test (SS#.jpg). The image in my previous post file name for my test is FO4-new_perk_shared_content.png

If it gives better result to have a transition state (even if only in memory), which I assume would crop the padding transparent area, I'm all for it.

> What are you expecting?

From my previous post: "I made a sample of 4 screenshots, the 2 middle close to what I want, the 3rd, being an exact match I'm looking for; so it would be something like this: 1 not even close, 2 pretty close, 3 pretty darn close / exact, 4 close to nothing matches."
If you need even more "visualization", it could be an example like (leading # matches the screenshot #) 1 1%, 2 75%, 3 95%, 4 0%, it's just an example and I have no idea what's the real #. The most important thing is that SS3.jpg has a very high match ratio.

**Reply by fmw42** [reply]
To compare and ignore transparent regions, you need to include a mask and be using IM 7. See https://imagemagick.org/discourse-server/viewtopic.php?f=4&t=31053

**Comment by DynV** [comment]
For the version see
> USER_NAME@SYSTEM_NAME ~
> $ compare --version
> Version: ImageMagick 7.0.10-27 Q16 x86_64 2020-08-14 https://imagemagick.org
> Copyright: © 1999-2020 ImageMagick Studio LLC
> License: https://imagemagick.org/script/license.php
> Features: Cipher DPC HDRI OpenMP(4.5)
> Delegates (built-in): bzlib cairo fftw fontconfig fpx freetype gslib heic jbig jng jp2 jpeg lcms lzma pangocairo png ps rsvg tiff webp x xml zlib

I don't need to ignore the transparent areas, I was trying to convey that if it make it easier to do the analysis, all the better. Also the same non-transparent area would need to be cropped on the comparison image and the compared image, something like:
1. get non-transparent area from comparison image
2. crop previous point area from copies of comparison & compared images
3. do the analysis of the 2 resulting images of the previous point.

I still don't know how to do what I'm looking for, what's below the quote in my previous post; if you would kindly help me with that.

**Reply by fmw42** [reply]
It would be better to use a mask from the alpha channel of the transparent image, then remove the alpha channels from any of the images and do the magick compare in IM 7 syntax as per the link above.

Post your transparent template image and one or two corresponding sized images to compare. I will help with the examples.

**Reply by DynV** [reply]
My transparent template image is in the post which you mentioned having a difficult time seeing, [here](https://user-images.githubusercontent.com/2823807/91781721-013dc280-ebc9-11ea-851d-1a99015b5884.png)'s a link to it. Please pick the corresponding sized images to compare from the link to where I uploaded SS1.jpg to SS4.jpg, which a 2nd link to is [here](https://imgur.com/a/Je4Ppp9). Again: The most important thing is that SS3.jpg has a very high match ratio.

**Comment by fmw42** [comment]
If I use -metric AE, I get identical values, because the green colors do not match perfectly (possibly due to jpg compression). So all the non-black pixels in the reference image (corresponding to the mask) have different colors than in the other two images. So the count corresponds to all the green in the reference.

```
magick 91781721-013dc280-ebc9-11ea-851d-1a99015b5884.png -alpha extract mask.png
magick 91781721-013dc280-ebc9-11ea-851d-1a99015b5884.png -alpha off reference.png

```


```
magick compare -metric AE -read-mask mask.png reference.png SS1.jpg null:
9431

magick compare -metric AE -read-mask mask.png reference.png SS4.jpg null:
9431

magick compare -metric AE -read-mask mask.png reference.png SS3.jpg null:
9370

```
However, if I use -metric rmse, then there is a difference.

```
magick compare -metric rmse -read-mask mask.png reference.png SS1.jpg null:
30654 (0.46775)

magick compare -metric rmse -read-mask mask.png reference.png SS4.jpg null:
31432.9 (0.479635)

magick compare -metric rmse -read-maskk.png reference.png SS3.jpg null:
2400.27 (0.0366257)

```

**Comment by DynV** [comment]
"The most important thing is that SS3.jpg has a very high match ratio."
screenshot THREE jpeg extension
I'd like you to ensure that both
1. SS3.jpg is a very close match, and
2. either SS1 or SS4 are a low / very low match.

is it the case for what you suggested?

**Reply by fmw42** [reply]
I added SS3.jpg to my post. AE shows the same result, since green colors are not perfectly matched. RMSE shows a much smaller (better match) result (about 4% different for SS3 rather than 47% different for SS1 and SS4 But since your images are JPG, the compression changes the colors so that you do not get a good match. You should save your images as PNG if possible rather than JPG. Do not convert the JPG to PNG as the compression will already be there changing your colors. Alternately, you might try using -remap to reduce colors to the same colormap before matching. Or alternately, convert your images to (binary) edges  before matching. That might give more distinctive results. 

Since you have JPG compression, another way with AE is to use a -fuzz value for tolerance to color change. This would be the simplest way to handle your images.

```
magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png SS1.jpg null:
8680

magick compare -fuzz 10% -metric AE -readmask mask.png reference.png SS4.jpg null:
8880

magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png SS3.jpg null:
0

```

**Reply by DynV** [reply]
Apologies about the delay. First, all screenshots are JPEGs and I created the mostly transparent PNG from those. I tried what you proposed with for result
> USER_NAME@SYSTEM_NAME ~
> $ magick FO4-new_perk_shared_content.png -alpha extract mask.png
> magick: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `mask.png' @ warning/png.c/MagickPNGWarningHandler/1748.
> 
> USER_NAME@SYSTEM_NAME ~
> $ magick FO4-new_perk_shared_content.png -alpha off reference.png
> 
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -metric rmse -read-mask mask.png reference.png SS/* null:
> 30606 (0.467018)
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -metric rmse -read-mask mask.png reference.png SS/SS
> SS1.jpg  SS2.jpg  SS3.jpg  SS4.jpg
> 
> USER_NAME@SYSTEM_NAME ~
> $ for i in SS/*; do magick compare -metric rmse -read-mask mask.png reference.png "$i"; done
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> 
> USER_NAME@SYSTEM_NAME ~
> $ for i in SS/*; do magick compare -metric rmse -read-mask mask.png reference.png SS/"$i"; done
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> compare:  `reference.png' @ error/compare.c/CompareImagesCommand/1150.
> 
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -metric rmse -read-mask mask.png reference.png SS/SS1.jpg null:
> 30606 (0.467018)
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -metric rmse -read-mask mask.png reference.png SS/SS2.jpg null:
> 15180.6 (0.231641)
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -metric rmse -read-mask mask.png reference.png SS/SS3.jpg null:
> 53.4298 (0.000815287)
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -metric rmse -read-mask mask.png reference.png SS/SS4.jpg null:
> 31402.7 (0.479174)
> 
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png SS/SS1.jpg null:
> 8728
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png SS/SS2.jpg null:
> 2963
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png SS/SS3.jpg null:
> 0
> USER_NAME@SYSTEM_NAME ~
> $ magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png SS/SS4.jpg null:
> 8911

and it works adequately beside my ineptitude at applying the command on all the files in a directory. With the 1st command there was a warning, but at least it wasn't an error.

I'm quite the novice at ImageMagick. Is the alternatives that you gave in your previous post ~mid 1st paragraph significantly better? This is for a 1-time use on hundreds of JPEG screenshots.

**Reply by fmw42** [reply]
You cannot use wild cards for image names in compare. It only compares two images at a time. You will need to write a loop over your input images and run the compare command in the script loop.  What is your OS -- Windows or Unix-like?

The warning is that a color profile was trying to be attached (automatically) for a grayscale image in the PNG reference. To avoid seeing warnings add -quiet to your command right after magick or after magick compare, etc.

I am not sure what you are asking about in the 1st paragraph. I suspect that the -fuzz approach is going to be the best method for you.

**Reply by DynV** [reply]
I'm using Cygwin.

**Reply by fmw42** [reply]
```
cd path_to/your_image_directory
for img in *.jpg; do
magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png "$img" null:
done

```

**Reply by DynV** [reply]
I ended up using the command
> for f in *.jpg; do i=${f##*/} i=${i%_*}; if (( 10#$i >= 20200717160912 )); then printf "%s\t" "$f"; /cygdrive/C/Program\ Files/ImageMagick-7.0.10-Q16-HDRI/magick.exe compare -fuzz 10% -metric AE -read-mask mask.png reference.png "$f" null:; printf "\n"; fi; done > FO4-Loriya-new_perk_comparison-200906-compare_fuzz_10pc_metric_AE.txt 2>&1

and I'm not sure if it got all of it, but 14 new perks out of 85 perk selection seems right; one may see what the FO4 PC looked like at L82 [here](https://imgur.com/a/7E75rqP).

**Update 1:**

Actually looking at that image there's 16 new perk and between L82 & 85 there was an additional one. It might have been because I forgot to screenshot new perks, but I doubt it since it's always something exciting to me, and I have a almost a reflex of taking screenshot exciting moments.

To get the 14 results, I cropped out everything that was 1000 or above.

**Update 2**

I broaden the results by taking anything below 3000. It turns out that every single result below 3000 down to 1000 have a timestamp close to a result below 1000 (which is as expected since I usually take a screenshot of the dialogue asking to confirm then another right after once it's done). Therefore I seriously doubt anything I found in this update is what I'm looking for (which should have a result under 1000).

**Comment by DynV** [comment, marked as answer]
I can't seem to mark a reply as an answer so the following refers to the reply in question, although I had to tweak it a bit which result is in [my reply to it](https://github.com/ImageMagick/ImageMagick/discussions/2490#discussioncomment-63468).
```
cd path_to/your_image_directory
for img in *.jpg; do
magick compare -fuzz 10% -metric AE -read-mask mask.png reference.png "$img" null:
done
```
_Originally posted by @fmw42 in https://github.com/ImageMagick/ImageMagick/discussions/2490#discussioncomment-63027_
